File: benchmarks/rsa_benchmark.py
from BCPython.BCWrapper import get_gateway
import timeit
import random


key_sizes = [1024, 2048, 3072, 4096]

gw = get_gateway()

RSAKeyPairGenerator = gw.jvm.org.bouncycastle.crypto.generators.RSAKeyPairGenerator
RSAKeyGenerationParameters = gw.jvm.org.bouncycastle.crypto.params.RSAKeyGenerationParameters
SecureRandom = gw.jvm.java.security.SecureRandom
RSAEngine = gw.jvm.org.bouncycastle.crypto.engines.RSAEngine
BigInteger = gw.jvm.java.math.BigInteger
Arrays = gw.jvm.java.util.Arrays

iters = 1000

for key_size in key_sizes:
    print(f"\n=== Тестирование RSA {key_size} бит ===")
    
    # RSA key gen
    generator = RSAKeyPairGenerator()
    generator.init(RSAKeyGenerationParameters(
        BigInteger("65537"),
        SecureRandom(),
        key_size,
        100  # certainty
    ))
    
    keyPair = generator.generateKeyPair()
    publicKey = keyPair.getPublic()
    privateKey = keyPair.getPrivate()
    
    # init engines
    encrypt_engine = RSAEngine()
    decrypt_engine = RSAEngine()
    
    # preparation of test data
    max_data_size = (key_size // 8) - 11
    test_data_size = min(32, max_data_size)
    
    # test data as Python bytes
    java_test_data = bytes([random.randint(0, 255) for _ in range(test_data_size)])
    
    
    print(f"Тест шифрования (данные: {test_data_size} байт):")
    
    def rsa_encrypt():
        encrypt_engine.init(True, publicKey)
        return encrypt_engine.processBlock(java_test_data, 0, len(java_test_data))
    
    for _ in range(3):
        timeit.timeit(rsa_encrypt, number=100)
    
    encrypt_times = []
    for _ in range(5):
        time_taken = timeit.timeit(rsa_encrypt, number=iters)
        encrypt_times.append(time_taken / iters)
    
    avg_encrypt = sum(encrypt_times) / len(encrypt_times)
    print(f"  Среднее время: {avg_encrypt * 1_000_000:.2f} мкс на операцию")
    
    encrypt_engine.init(True, publicKey)
    encrypted_data = encrypt_engine.processBlock(java_test_data, 0, len(java_test_data))
    

    print(f"Тест расшифрования:")
    
    def rsa_decrypt():
        decrypt_engine.init(False, privateKey)
        return decrypt_engine.processBlock(encrypted_data, 0, len(encrypted_data))
    
    for _ in range(3):
        timeit.timeit(rsa_decrypt, number=100)
    
    decrypt_times = []
    for _ in range(5):
        time_taken = timeit.timeit(rsa_decrypt, number=iters)
        decrypt_times.append(time_taken / iters)
    
    avg_decrypt = sum(decrypt_times) / len(decrypt_times)
    print(f"  Среднее время: {avg_decrypt * 1_000_000:.2f} мкс на операцию")
    

    print(f"Тест полного цикла (шифрование + расшифрование):")
    
    def rsa_full_cycle():
        encrypt_engine.init(True, publicKey)
        encrypted = encrypt_engine.processBlock(java_test_data, 0, len(java_test_data))
        
        decrypt_engine.init(False, privateKey)
        return decrypt_engine.processBlock(encrypted, 0, len(encrypted))
    
    for _ in range(3):
        timeit.timeit(rsa_full_cycle, number=50)
    
    cycle_times = []
    for _ in range(5):
        time_taken = timeit.timeit(rsa_full_cycle, number=iters // 2)
        cycle_times.append(time_taken / (iters // 2))
    
    avg_cycle = sum(cycle_times) / len(cycle_times)
    print(f"  Среднее время: {avg_cycle * 1_000_000:.2f} мкс на цикл")

File: benchmarks/src/main/java/benchmarks/RSABenchmark.java
package benchmarks;

import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.engines.RSAEngine;
import org.bouncycastle.crypto.generators.RSAKeyPairGenerator;
import org.bouncycastle.crypto.params.*;
import org.openjdk.jmh.annotations.*;

import java.security.SecureRandom;
import java.util.concurrent.TimeUnit;
import java.math.BigInteger;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@State(Scope.Thread)
@Fork(0)
public class RSABenchmark {

    @Param({"1024", "2048", "3072", "4096"})
    public int keySize;
    
    private AsymmetricCipherKeyPair keyPair;
    private RSAEngine encryptEngine;
    private RSAEngine decryptEngine;
    private byte[] testData;
    private byte[] encryptedData;

    @Setup(Level.Trial)
    public void setup() {
        // generation of RSA keys
        RSAKeyPairGenerator generator = new RSAKeyPairGenerator();
        generator.init(new RSAKeyGenerationParameters(
            BigInteger.valueOf(65537),
            new SecureRandom(),
            keySize,
            100 // certainty
        ));
        
        keyPair = generator.generateKeyPair();
        
        // init engines
        encryptEngine = new RSAEngine();
        encryptEngine.init(true, keyPair.getPublic());
        
        decryptEngine = new RSAEngine();
        decryptEngine.init(false, keyPair.getPrivate());
        
        // preparation of test data
        int maxDataSize = (keySize / 8) - 11; // for PKCS#1 v1.5 padding
        testData = new byte[Math.min(32, maxDataSize)]; // 32 bytes or less
        new SecureRandom().nextBytes(testData);
        
        // pre-encryption of data for decryption tests
        encryptedData = encryptEngine.processBlock(testData, 0, testData.length);
    }

    @Benchmark
    public byte[] rsa_encrypt() {
        encryptEngine.init(true, keyPair.getPublic());
        return encryptEngine.processBlock(testData, 0, testData.length);
    }

    @Benchmark
    public byte[] rsa_decrypt() {
        decryptEngine.init(false, keyPair.getPrivate());
        return decryptEngine.processBlock(encryptedData, 0, encryptedData.length);
    }

    @Benchmark
    public byte[] rsa_full_cycle() {
        encryptEngine.init(true, keyPair.getPublic());
        byte[] encrypted = encryptEngine.processBlock(testData, 0, testData.length);
        
        decryptEngine.init(false, keyPair.getPrivate());
        return decryptEngine.processBlock(encrypted, 0, encrypted.length);
    }

    public static void main(String[] args) throws Exception {
        org.openjdk.jmh.Main.main(args);
    }
}